### PR TITLE
[ExpressionLanguage] Add a way to hook on each node when dumping the AST

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | "master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19013 |
| License | MIT |
| Doc PR | - |

This is an iteration over #19013 to allow writing dumpers that can decorate dumps (e.g. add HTML tags based on each node type to do syntax highlighting).

### CODE REVIEW

File: src/Symfony/Component/ExpressionLanguage/Parser.php [hunk 1]
```diff
@@ -330,7 +330,7 @@ public function parsePostfixExpression($node)
                     throw new SyntaxError('Expected name', $token->cursor);
                 }
 
-                $arg = new Node\ConstantNode($token->value);
+                $arg = new Node\NameNode($token->value);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: Using `NameNode` instead of `ConstantNode` for representing `GetAttrNode::PROPERTY_CALL` and `GetAttrNode::METHOD_CALL` (but keep `ConstantNode`for `GetAttrNode::ARRAY_CALL`), the AST is more correct.
This is enlighten by the "unescaping" ([this trim](https://github.com/symfony/symfony/pull/19060/files#diff-ba72ffdc5e1ef963bfb7072610580259L102)) that was required otherwise when dumping a `GetAttrNode`.

File: src/Symfony/Component/ExpressionLanguage/ParsedExpression.php [hunk 2]
```diff
@@ -42,6 +42,17 @@ public function getNodes()
 
     public function dump()
     {
-        return $this->nodes->dump();
+        return $this->dumpNode($this->nodes);
+    }
+
+    protected function dumpNode(Node $node)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/ParsedExpression.php]: `protected` on purpose since this is exactly the extension point that is missing currently to hook into dumping each node

File: src/Symfony/Component/ExpressionLanguage/Tests/Node/ConstantNodeTest.php [hunk 3]
```diff
@@ -47,8 +47,8 @@ public function getDumpData()
             array('false', new ConstantNode(false)),
             array('true', new ConstantNode(true)),
             array('null', new ConstantNode(null)),
-            array(3, new ConstantNode(3)),
-            array(3.3, new ConstantNode(3.3)),
+            array('3', new ConstantNode(3)),
```
Comment (lyrixx) [inline src/Symfony/Component/ExpressionLanguage/Tests/Node/ConstantNodeTest.php]: that's wrong.

Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Tests/Node/ConstantNodeTest.php]: how that? a dump is a string, this is now always the case

File: src/Symfony/Component/ExpressionLanguage/ParsedExpression.php (hunk 2 above)
Comment (lyrixx) [inline src/Symfony/Component/ExpressionLanguage/ParsedExpression.php]: IMHO, the extension point is not really easy to use. Because the EL returns a `ParsedExpression` and not a `ParsedExpressionToDumpInAnHTMLShape` you have to do some ugly code.

File: src/Symfony/Component/ExpressionLanguage/Tests/Node/ConstantNodeTest.php (hunk 3 above)
Comment (lyrixx) [inline src/Symfony/Component/ExpressionLanguage/Tests/Node/ConstantNodeTest.php]: got it ;)

File: src/Symfony/Component/ExpressionLanguage/Node/ArgumentsNode.php [hunk 4]
```diff
@@ -27,12 +27,14 @@ public function compile(Compiler $compiler)
...
     {
-        $str = '';
+        $tokens = array();
 
         foreach ($this->getKeyValuePairs() as $pair) {
-            $str .= sprintf('%s, ', $pair['value']->dump());
+            $tokens[] = ', ';
+            $tokens[] = $pair['value'];
         }
+        array_shift($tokens);
```
Comment (stof) [inline src/Symfony/Component/ExpressionLanguage/Node/ArgumentsNode.php]: isn't it better to add the ``,  `` token after each pair, and to pop the last one here ? It would avoid having to reindex the array

File: src/Symfony/Component/ExpressionLanguage/ParsedExpression.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Component/ExpressionLanguage/ParsedExpression.php]: the right extension point is probably to have this ``dumpNode`` method in a separate class, allowing to use a different implementation

File: src/Symfony/Component/ExpressionLanguage/Node/ArgumentsNode.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/Node/ArgumentsNode.php]: fixed

File: src/Symfony/Component/ExpressionLanguage/ParsedExpression.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/ParsedExpression.php]: True! I now made the method private. There is no need for any extension point: one should just call $expr->getNodes() then do things with the result.

Comment (lyrixx) [inline src/Symfony/Component/ExpressionLanguage/ParsedExpression.php]: I agree with @stof ; It was my first idea when I implemented this.

Comment (nicolas-grekas) [inline src/Symfony/Component/ExpressionLanguage/ParsedExpression.php]: could be done in another PR

File: src/Symfony/Component/ExpressionLanguage/Node/ArgumentsNode.php [hunk 5]
```diff
@@ -27,12 +27,14 @@ public function compile(Compiler $compiler)
 
     public function dump()
     {
-        $str = '';
+        $tokens = array();
```
Comment (fabpot) [inline src/Symfony/Component/ExpressionLanguage/Node/ArgumentsNode.php]: Token has another meaning in the component, so we need to use something else.

File: src/Symfony/Component/ExpressionLanguage/Parser.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: I tend to agree that this semantically more correct. Is it a BC break? I don't think so.

**Comment by fabpot** [comment]
You should also add a phpdoc to the `dump()` method on `Node` at least for the return value.

**Comment by fabpot** [comment]
And I think that the `dump()` method on `Node`s should be changed to something else as I would expect a dump method to return a string, not an array.

**Comment by nicolas-grekas** [comment]
Comments addressed, `dump()` renamed to `toArray()`

**Comment by fabpot** [comment]
👍

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Component/ExpressionLanguage/Parser.php (hunk 1 above)
Comment (lyrixx) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: Yes it's a BC break. I created [a melody script](https://gist.github.com/lyrixx/d9e7f17c86d1ed8ae3869f23bbb254c8) to reproduce it.

@fabpot it's because of this change your gist your sent me yesterday does not work anymore.

Comment (fabpot) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: Arf, any change is a BC break after all :)

Comment (fabpot) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: So, we need to revert this change, that's two issues as of now.

Comment (fabpot) [inline src/Symfony/Component/ExpressionLanguage/Parser.php]: reverted now
